### PR TITLE
feat(orc8r): Create SyncSubscribers logic for subscriberdb servicer

### PR DESCRIPTION
**Pull request by wangyyt1013**
Signed-off-by: Yuanyuting Wang <yw3241@columbia.edu>

## Summary

1. Add implementation for `subscriberdb/SyncSubscribers` rpc endpoint
2. Return digests for `subscriberdb/ListSubscribers`
3. Adapt agw to changes in `ListSubscribers` so as not to be backward-breaking

## Test Plan

1. Ran all existing tests; all passed
2. Added appropriate tests for newly added features

## Additional Information

- This PR is based on #7810

### CODE REVIEW

**Review by themarwhal — APPROVED** [review]

File: lte/protos/subscriberdb.proto [hunk 1]
```diff
@@ -240,11 +240,11 @@ message SyncSubscribersResponse {
   // per_sub_digests is a list of digests for each server-side subscriber, ordered
   // by their IDs (IMSI).
   repeated SubscriberDigestWithID per_sub_digests = 3;
-  // added_modified is the set of all subscriber data modified/added on the server
-  // side, indexed by their IDs (IMSI).
-  map<string, SubscriberData> added_modified = 4;
-  // removed is a list of IDs for subscribers removed on the server side.
-  repeated string removed = 5;
+  // deleted is the set of all subscriber data modified/added on the server side,
```
Comment (hcgatewood) [inline lte/protos/subscriberdb.proto]: "to_renew is..." ?

File: lte/cloud/configs/subscriberdb.yml [hunk 2]
```diff
@@ -12,3 +12,6 @@
 
 # flatDigestEnabled is a feature flag for the flat digest functionality.
 flatDigestEnabled: true
+# maxNoResyncChangesetSize specifies the max size of the cloud-agw changeset
+# past which a resync signal will be sent back to the agw.
+maxNoResyncChangesetSize: 500
```
Comment (hcgatewood) [inline lte/cloud/configs/subscriberdb.yml]: Maybe "changesetSizeTheshold"

File: lte/cloud/go/services/subscriberdb/servicers/subscriberdb_servicer.go [hunk 3]
```diff
@@ -107,11 +144,21 @@ func (s *subscriberdbServicer) ListSubscribers(ctx context.Context, req *lte_pro
 		return nil, err
 	}
 
-	digest, _ := s.getDigestInfo(&lte_protos.Digest{Md5Base64Digest: ""}, networkID)
+	flatDigest, perSubDigests := &lte_protos.Digest{Md5Base64Digest: ""}, []*lte_protos.SubscriberDigestWithID{}
```
Comment (hcgatewood) [inline lte/cloud/go/services/subscriberdb/servicers/subscriberdb_servicer.go]: Move these declarations to separate lines

File: lte/cloud/go/services/subscriberdb/servicers/subscriberdb_servicer.go [hunk 4]
```diff
@@ -107,11 +144,21 @@ func (s *subscriberdbServicer) ListSubscribers(ctx context.Context, req *lte_pro
...
+	if req.PageToken == "" {
+		// the digests are sent back during the request for the first page of subscriber data
```
Comment (hcgatewood) [inline lte/cloud/go/services/subscriberdb/servicers/subscriberdb_servicer.go]: Sentence-case your comments

Comment (hcgatewood) [inline lte/cloud/go/services/subscriberdb/servicers/subscriberdb_servicer.go]: Move this comment above the if-block (swap with line above)

File: lte/cloud/go/services/subscriberdb/digest.go [hunk 5]
```diff
@@ -107,27 +107,29 @@ func GetPerSubscriberDigestsDiff(prev []*lte_protos.SubscriberDigestWithID, next
 	deleted := []string{}
 
 	for i < n && j < m {
-		iSid, jSid := prev[i].Sid.Id, next[j].Sid.Id
-		if iSid == jSid {
+		iSidStr, jSidStr := lte_protos.SidString(prev[i].Sid), lte_protos.SidString(next[j].Sid)
```
Comment (hcgatewood) [inline lte/cloud/go/services/subscriberdb/digest.go]: Why the change?

File: lte/cloud/go/services/subscriberdb/digest_test.go [hunk 6]
```diff
@@ -384,50 +384,50 @@ func TestGetPerSubscriberDigestsDiff(t *testing.T) {
...
 		toRenew, deleted := subscriberdb.GetPerSubscriberDigestsDiff(prev, next)
 		assert.Equal(t, expectedToRenew, toRenew)
 		assert.Empty(t, deleted)
 	})
 	t.Run("both not empty, basic", func(t *testing.T) {
 		prev := []*lte_protos.SubscriberDigestWithID{
-			{Sid: &lte_protos.SubscriberID{Id: "00001"}, Digest: &lte_protos.Digest{Md5Base64Digest: "apple"}},
-			{Sid: &lte_protos.SubscriberID{Id: "00002"}, Digest: &lte_protos.Digest{Md5Base64Digest: "banana"}},
-			{Sid: &lte_protos.SubscriberID{Id: "00004"}, Digest: &lte_protos.Digest{Md5Base64Digest: "dragonfruit"}},
+			{Sid: &lte_protos.SubscriberID{Id: "00001", Type: lte_protos.SubscriberID_IMSI}, Digest: &lte_protos.Digest{Md5Base64Digest: "apple"}},
```
Comment (hcgatewood) [inline lte/cloud/go/services/subscriberdb/digest_test.go:415]: Yeah technically IMSIs need to fit this regex (e.g. `IMSI0000000000042`), but it's not a huge deal for these types of tests https://github.com/magma/magma/blob/ea55e13ba409a868a7fa27220d3a6ae89238d2ca/orc8r/cloud/go/obsidian/swagger/v1/swagger.yml#L11425-L11429

File: lte/cloud/go/services/subscriberdb/load.go [hunk 7]
```diff
@@ -22,8 +22,10 @@ import (
 	lte_models "magma/lte/cloud/go/services/lte/obsidian/models"
 	"magma/lte/cloud/go/services/subscriberdb/obsidian/models"
 	"magma/orc8r/cloud/go/services/configurator"
+	storage2 "magma/orc8r/cloud/go/storage"
```
Comment (hcgatewood) [inline lte/cloud/go/services/subscriberdb/load.go]: No need to alias this

**Review by hcgatewood — APPROVED** [review]
Lgtm! Couple small nits / requests for clarifications, then msg me to merge

**Comment by hcgatewood** [comment]
Test comment for checking Toast integration

File: lte/cloud/go/services/subscriberdb/digest.go (hunk 5 above)
Comment (wangyyt1013) [reply lte/cloud/go/services/subscriberdb/digest.go]: Through this change, the entire string of `Sid` will be sent down to agw and stored in the agw database, instead of only the `Sid.Id` (e.g. "IMSI00000" vs. "00000") @hcgatewood

File: lte/cloud/go/services/subscriberdb/digest_test.go (hunk 6 above)
Comment (wangyyt1013) [reply lte/cloud/go/services/subscriberdb/digest_test.go:415]: Yeah this is kinda a nit, just wanted to change these to 5 digits for consistency (since all other mock IMSIs in this file are 5 digits) @hcgatewood

File: lte/cloud/go/services/subscriberdb/digest.go (hunk 5 above)
Comment (hcgatewood) [reply lte/cloud/go/services/subscriberdb/digest.go]: Gotcha. Didn't realize that's how it worked. So I like the content of the change, but just keep the var names as `iSid` and `jSid` for conciseness

**Review by ardzoht — APPROVED** [review]
lgtm
